queasars/job_shop_scheduling/scip_solver.py

```diff
@@ -0,0 +1,136 @@
+                scheduled_operations.append(ScheduledOperation(operation=operation, start=start_time))
+            job_schedules[job] = tuple(scheduled_operations)
+
+        return JobShopSchedulingResult(problem_instance=self._jssp_instance, schedule=job_schedules)
+
+    def _prepare_variables(self):
+        """
+        Creates the integer variables for the starting times of each operation and constrains them to be at least 0.
+        Also creates a integer variable to hold the makespan, which is constrained to be larger than the end times
+        of the last operations. Sets the optimization goal to minimize the makespan variable
+        """
+        # create auxiliary variable to hold the makespan as an optimization goal
+        self._optimization_var = self._model.addVar("optimization_var", vtype="INTEGER")
+        self._model.setObjective(self._optimization_var)
+        self._model.addCons(self._optimization_var >= 0)
+
+        for job in self._jssp_instance.jobs:
+            for operation in job.operations:
+                # Cache the operations per machine
+                # This is needed for the additional constraints
+                if operation.machine not in self._machine_operations:
+                    self._machine_operations[operation.machine] = []
+                self._machine_operations[operation.machine].append(operation)
+
+                # Create an integer variable for the start time of each operation
+                # Constrain it to be at least 0
+                var = self._model.addVar(operation.identifier, vtype="INTEGER")
+                self._operation_start_variables[operation] = var
+                self._model.addCons(var >= 0)
+
+    def _prepare_constraints(self):
+        """
+        Constrains all operations within each job to start in order and not overlap.
+        Constrains all operations for each machine to not overlap. To this end auxiliary binary variables
+        are created to enable ordering decisions.
+        """
+        for job in self._jssp_instance.jobs:
+            n_operations = len(job.operations)
+
+            # Ensure by constraint, that each operation in a job finishes, before the next one may start
+            for i in range(0, n_operations - 1):
+                self._model.addCons(
+                    self._operation_start_variables[job.operations[i]] + job.operations[i].processing_duration
+                    <= self._operation_start_variables[job.operations[i + 1]]
+                )
+            # For each job the makespan must be larger than the end time of its last operation
+            self._model.addCons(
+                self._operation_start_variables[job.operations[n_operations - 1]]
+                + job.operations[n_operations - 1].processing_duration
+                <= self._optimization_var
+            )
+
+        # Ensure for each machine, that no operation may overlap
+        for operations in self._machine_operations.values():
```

Comment:
While correct, I would suggest iterating over `enumerate()` and then maybe putting the machine name into the constraint names below. Otherwise the variable name `operations` seems to contradict the comment on the line before (which it doesn't if one knows the type of `self._machine_operations`).

Reply:
I see. I changed this to: `for machine, operations in self._machine_operations.items()` i think this should be clearer in meaning.

Reply:
Perfect, thanks!